Add ability to edit run names in the UI

smurching opened the pull request:
Adds the ability to edit run names in the UI, based on mocks from @gioa.

## Screenshots
### Rename run dropdown
![screen shot 2018-09-05 at 3 38 29 pm](https://user-images.githubusercontent.com/2358483/45125263-a03a9b80-b122-11e8-8c21-71cde3b57ae2.png)
![screen shot 2018-09-05 at 3 38 20 pm](https://user-images.githubusercontent.com/2358483/45125262-9fa20500-b122-11e8-9706-8bea691cff37.png)
### Rename run modal & validation
![screen shot 2018-09-05 at 3 43 14 pm](https://user-images.githubusercontent.com/2358483/45125264-a03a9b80-b122-11e8-8f03-4f8de7c64241.png)
![screen shot 2018-09-05 at 3 43 31 pm](https://user-images.githubusercontent.com/2358483/45125265-a03a9b80-b122-11e8-8279-8ddea93d70bc.png)
### Handling server errors
If a request to the server fails, we show the server error inline in the modal - we expect this to happen only rarely.
![screen shot 2018-09-05 at 3 43 42 pm](https://user-images.githubusercontent.com/2358483/45125266-a03a9b80-b122-11e8-803f-a5df1355035f.png)

## Implementation
Python API changes:
* Specify run name by setting an "_MLFLOW_RUN_NAME" tag when creating a run
* Pass an empty string for the `run_name` field of `RunInfo`, to prepare for deprecating the `run_name` field.

JS changes:
* Adds action / reducer logic for setting tags
* Adds a `RenameRunFormView` containing the rename-run form
* Adds a `RenameRunModal` that renders a `RenameRunFormView`, passing down the callbacks needed to submit data to the server / close the parent modal on form submission
* Adds a dropdown menu in `RunView` containing a "Rename Run" option that, when clicked, renders a `RenameRunModal`. Other items (e.g. "Delete Run") could be added to the menu moving forward.
* Updates to `BreadcrumbTitle` & `CompareRunView` to use the new tag for rendering the run name

## Testing
Manually verified the following behaviors:
* Renaming runs works in the typical case
* Users can't close the rename dialog while the form is submitting (simulated by adding a sleep to the server)
* Failed rename-run form submissions result in an alert --> redirect to the parent experiment page

TODO: Probably need to change the Java API in the same way as the Python one. Also, IMO the relationship between RenameRunModal & RenameRunFormView is a bit complicated (the form performs lots of operations that affect the modal), would appreciate any suggestions towards simplifying that.

smurching committed:
Show run name in UI if set, otherwise retain current behavior of
showing 'Run <uuid>'

smurching committed:
WIP: Made run name editable via the UI, TODO:
* Sync on right UX for editing run name
* Modify CreateRun API to no longer pass run name (probably: still take a run name, but stop setting the run name field in the DB)

smurching committed:
WIP updating the compare run view

mateiz reviewed (commented):
FYI this didn't work for me. It seemed to have passed the empty string as a tag name to the server, because I saw an error about that in the `mlflow ui` logs.

In terms of UX, I think we should replace it with an edit button next to the name at the top of the page. That's the more obvious thing to click to edit it.

smurching committed:
Bugfix to input form approach. About to refactor to use dialog

smurching committed:
Updates, about to add dropdown

smurching committed:
WIP

smurching committed:
WIP

smurching committed:
WIP: Copy in a bubnch of stuff from ui component example

smurching committed:
WIP, modal shows up

smurching committed:
Updates, renaming works via a button now

smurching committed:
Add x & esc-key handling to modal

smurching committed:
Merge branch 'master' into run-name-updates

smurching committed:
Remove unused formik files

smurching committed:
WIP styling. TODO: handle errors, handle compare run

smurching committed:
WIP working on updating CompareRunView. Might move updates into a follow-up

smurching committed:
WIP adding error-handling using modals. Will instead
wait to add handling once error page is ready

smurching committed:
WIP styling

smurching committed:
WIP styling some more. TODO: simplify modal (remove container), disable escaping modal via click

smurching committed:
WIP trying ReactModal

smurching committed:
More WIP, going to just disable closing modal when clicking overlay

smurching committed:
Move onClose logic to form to avoid cleaning up form component after it's unmounted by
removing the modal

smurching committed:
Remove ModalsContainer, clean up some other things

smurching committed:
Try changing Python client - stop passing run ID to REST API, instead start setting it via a tag.
This will break backwards compatibility with new MLflow clients & old servers and UIs, in that old servers will
no longer receive a run name from new clients. However no released MLflow UI shows the run name, so we shouldn't
run into big issues

smurching committed:
WIP

smurching committed:
Modal styling

smurching committed:
Merge master

smurching committed:
Code cleanups

smurching committed:
Code cleanup

smurching commented on mlflow/entities/run_tag.py:
```diff
@@ -4,6 +4,8 @@
 
 class RunTag(_MLflowObject):
     """RunTag object for Python client."""
+    _RUN_NAME_TAG_NAME = "_MLFLOW_RUN_NAME"
```
Ah thanks for the pointer - how do you feel about e.g. "mlflow.entities.run.runName"?

smurching committed:
Small simplifying changes (narrow some internal JS helpers' APIs, etc)

smurching committed:
Fix lint

smurching committed:
More cleanups

smurching committed:
WIP

smurching committed:
WIP

andrewmchen commented on mlflow/server/js/src/components/MetricPage.js:
```diff
@@ -26,6 +26,10 @@ class MetricPage extends Component {
       const requestId = getUUID();
```
nit: more descriptive name for htis

andrewmchen commented on mlflow/server/js/src/components/MetricView.js:
```diff
@@ -101,11 +116,22 @@ const mapStateToProps = (state, ownProps) => {
   });
+  let runName
+  if (runUuids.length > 0) {
+    const tags = getRunTags(runUuids[0], state);
```
Could we remove this.

andrewmchen commented on mlflow/server/js/src/components/RunView.js:
```diff
@@ -83,8 +90,17 @@ class RunView extends Component {
+    this.setState({showRunRenameModal: true});
+  };
+
+  hideModal() {
```
nit: `hideModal` -> `hideRenameRunModal`

andrewmchen commented on mlflow/server/js/src/components/RunView.js:
```diff
@@ -83,8 +90,17 @@ class RunView extends Component {
   }
 
+  handleRenameRunClick() {
+    this.setState({showRunRenameModal: true});
```
nit: add spaces after and before `{` and `}`

andrewmchen commented on mlflow/server/js/src/components/RunView.js:
```diff
@@ -233,7 +267,9 @@ const mapStateToProps = (state, ownProps) => {
   const latestMetrics = getLatestMetrics(runUuid, state);
-  return { run, experiment, params, tags, latestMetrics, metricPageRoute, getMetricPagePath };
+  const runName = Utils.getRunDisplayName(tags, runUuid);
+  debugger;
```
remove!

mateiz commented:
I just tried this out and saw a few small issues:
- The caret next to the name is too low on the page somehow and looks weird. It would be great to get Joy to look at it and tweak it.
- Instead of "Rename Run", should it just say Rename? Other similar menus say only Rename (e.g. a notebook's File menu).
- (Maybe for a later PR:) The names aren't shown on the Experiment page yet; we'll need to figure out where/how to show them since people will use that to search.

smurching committed:
WIP

smurching committed:
Replace inline styles with JS objects

smurching committed:
Code style

smurching committed:
Try to fix more JS lint issues

smurching committed:
Merge branch 'master' into run-name-updates

smurching committed:
Install eslint & fix lint errors locally

smurching committed:
Slightly tweak position of close button

smurching committed:
dropdown styling

smurching committed:
Update style based on comments from Joy

smurching committed:
Fix JS lint

mateiz commented on mlflow/server/js/src/components/RunView.js:
```diff
@@ -156,6 +193,10 @@ class RunView extends Component {
+          <div className="run-info">
+            <span className="metadata-header">Run ID: </span>
+            <span className="metadata-info">{run.getRunUuid()}</span>
+          </div>
```
Put this farther up in the list, after Date. It should be in a consistent place for all runs since people will often copy-paste it.

smurching committed:
Move run UUID higher up, show full git commit in run detail page, increase width of run view columns to properly align things when we show full git hash

smurching committed:
Fix lint

mateiz commented on mlflow/server/js/src/components/BreadcrumbTitle.js:
```diff
@@ -37,7 +38,7 @@ export default class BreadcrumbTitle extends Component {
     const chevron = <i className="fas fa-chevron-right breadcrumb-chevron" key="chevron"/>;
     return (
-      <h1>
+      <h1 style={styles.h1}>
```
Why did you use a `style` tag here instead of just putting it in the CSS file? I think we can have `display: inline` on all `h1` elements; I don't see why it would hurt.

mclin replied:
Changing h1 to `display: inline` I think is kind of unexpected. If you want something adjacent to this, make the parent `display: flex` (preferred, but need to double check in IE11), or use `float`.

smurching replied:
Ended up making the parent `display: flex`, making the `h1` have an inline style would work too but seemed like a larger change

mateiz commented on mlflow/server/js/src/components/RunView.js:
```diff
@@ -272,3 +313,15 @@ const shellEscape = (str) => {
   return str;
 };
 
+const styles = {
```
Put the styles in the CSS if possible, not here, since it's confusing to split them across two types of files.

mateiz commented on mlflow/server/js/src/components/modals/RenameRunFormView.js:
```diff
@@ -0,0 +1,82 @@
+            name="newRunName"
+            label="New Run Name"
+            autoFocus
+            style={{"width": "100%"}}
```
Nit: Is there any way we could make this open with the text all selected by default, so that you can just type in a new name? I think you can add `onfocus="this.select()"` on it.

mclin commented on mlflow/server/js/src/components/MetricView.js:
```diff
@@ -101,11 +116,16 @@ const mapStateToProps = (state, ownProps) => {
       metrics[i][runUuid] = entries[i].value;
     }
   });
+  const runNames = runUuids.map((runUuid) => {
```
Creating objects in `mapStateToProps` makes it so you can never use PureComponent to optimize rendering. Every time redux store updates, this will create a new `runNames` object (and `metrics` object), which will cause shallowEquals of props to be false, triggering a render. Going forward you might want to calculate derived data in a selector, so it can be memoized with [reselect](https://www.npmjs.com/package/reselect).

mclin commented on mlflow/server/js/src/components/modals/RenameRunFormView.js:
```diff
@@ -0,0 +1,82 @@
+            style={{"width": "100%"}}
+        />
+      </div>
+      <div style={styles.buttonsDiv}>
```
I wonder if it's possible to use the bootstrap modal styles here to save you some effort.

https://react-bootstrap.github.io/components/modal/

mclin commented on mlflow/server/js/src/components/modals/RenameRunFormView.js:
```diff
@@ -0,0 +1,82 @@
+
+const styles = {
+  form: {
+    width: "480px",
```
You can put a number as a shorthand for "px".
"480px" => 480

mclin commented on mlflow/server/js/src/components/modals/RenameRunFormView.js:
```diff
@@ -0,0 +1,82 @@
+    } = renderProps;
+    const saveText = isSubmitting ? "Saving..." : "Save";
+    return <form onSubmit={handleSubmit} style={styles.form}>
+      <div style={styles.formField}><h2 style={{"marginTop": "0px"}}>Rename Run</h2></div>
```
This is basically the modal header? It's a bit brittle to reuse styles.formField here. I wonder if it's possible to use bootstrap styles here.

mclin commented on mlflow/server/js/src/components/modals/RenameRunModal.js:
```diff
@@ -0,0 +1,118 @@
+      setTagApi(this.props.runUuid, tagKey, newRunName, setTagRequestId)).catch((err) => {
+        // TODO: remove alert, redirect to an error page on failed requests once one exists
+        // eslint-disable-next-line no-alert
+        alert("Unable to rename run, got error '" + err + "'. Redirecting to parent experiment " +
```
For the TODO, this seems a reasonable pattern for handling a failed submit (under `Describe alternatives you've considered`)
https://github.com/jaredpalmer/formik/issues/711

I wouldn't redirect on error, because it loses the user's input.

smurching replied:
Unfortunately for now the error message is a generic "Failed to rename run, error: XHR failed" but it seems possible to improve by adding logic [here](https://github.com/mlflow/mlflow/blob/4047ac4b81cf8d02f2a48a7e0cb0f6098ddcc8ed/mlflow/server/js/src/Actions.js#L103). One possibility is to at least propagate/render the server error code (e.g. `"Failed to rename run. Error message: " + errorMessage + ", server error code: " + errorCode"`) for debuggability

mclin commented on mlflow/server/js/src/components/modals/RenameRunModal.js:
```diff
@@ -0,0 +1,118 @@
+        "page.");
+        this.props.history.push(Routes.getExperimentPageRoute(this.props.experimentId));
+      }).finally(() => {
+        this.setState({ isSubmittingState: false });
```
It's a bit unfortunate to be duplicating this submitting state with Formik, but if you want to prevent closing the modal while submitting, I don't see a better way.

mclin reviewed (commented):
Nice! Yay, Formik! A few comments.

mclin commented on mlflow/server/js/src/components/modals/RenameRunFormView.js:
```diff
@@ -0,0 +1,82 @@
+    const saveText = isSubmitting ? "Saving..." : "Save";
+    return <form onSubmit={handleSubmit} style={styles.form}>
+      <div style={styles.formField}><h2 style={{"marginTop": "0px"}}>Rename Run</h2></div>
+      <div style={styles.label}>New run name:</div>
```
You should put the label inside the Field. You could use bootstrap styles as well. Something like this (had this lying around):
```
import { FormGroup, ControlLabel, FormControl, HelpBlock } from 'react-bootstrap';

export default function TextField(
  { field, form: { touched, errors }, label, ...props },
) {
  const error = errors[field.name];
  return <FormGroup
    controlId={field.name}
    validationState={touched[field.name] && error ? 'error' : null}
  >
    <ControlLabel>{label}</ControlLabel>
    <FormControl
      type="text"
      {...field}
    />
    { touched[field.name] && error ? <HelpBlock>{error}</HelpBlock> : null }
  </FormGroup>
}
```

smurching committed:
WIP

smurching committed:
WIP; trying to use modal styles in form

smurching committed:
Revert modal changes

smurching committed:
Address most review comments. TODO:
* Slightly better error handling?
* Try to consolidate CSS into fewer files
* Fix lint errors

smurching committed:
Update TextField to fix lint

smurching committed:
Move other styles into CSS files

mclin commented on mlflow/server/js/src/components/modals/TextField.js:
```diff
@@ -0,0 +1,38 @@
+
+import { FormGroup, ControlLabel, FormControl, HelpBlock } from 'react-bootstrap';
+
+class TextField extends Component {
```
I'd move this to a `components/fields` folder.

mclin commented on mlflow/server/js/src/components/modals/TextField.js:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  render() {
+    const field = this.props.field;
```
For this to be reusable, we need a way to pass props directly to the `<FormControl>`.

eg `const { form: {touched, errors},  label, field, ...props } = this.props;`

then pass `{...props}` after `{...field}` in the `FormControl`. That way you can customize `onFocus` from the outside.

mclin reviewed (commented):
Nice! One more pointer.

smurching committed:
Merge master

smurching committed:
Add XHR error-message handling, pass props to FormControl

smurching committed:
Remove autofocus from proptypes

smurching committed:
Remove unused & vaguely-named styles, since styles in .css files are
applied globally

smurching committed:
Address additional comments:
* Add back x
* Add space for error
* Move global error down
* Restore padding back to original values, remove line for footer

smurching committed:
Add modal fade-in & fade-out

mclin commented:
LGTM!

smurching committed:
Fix padding

smurching commented:
Sync'd offline with @gioa, @mclin, @andrewmchen, @mateiz. Merging, thanks all of you for the feedback & reviews!
